lib/Runtime/Library/Utf8String.h

```diff
@@ -0,0 +1,142 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+#pragma once
+
+#include "Codex/Utf8Codex.h"
+
+namespace Js
+{
+    class Utf8String : public JavascriptString
+    {
+    private:
+        typedef struct {
+            FieldNoBarrier(size_t) length;
+            Field(char*) buffer;
+        } PrefixedUtf8String;
+        Field(PrefixedUtf8String*) utf8String;
```

Comment:
If PrefixedUtf8String is only two pointers big, could we just put it directly in the object rather than requiring a separate allocation (and a separate pointer dereference every time we use it)?

Reply:
Part of the original motivation of this was to allow other kinds of strings to convert themselves into a `Utf8String`, and when I was doing that one of the candidate string types only had 1 pointer's worth of space available, which is why I went down this path (see the `ConvertString` method which implicitly assumes that there is space for the one pointer available). However right now that's not happening, so I could undo that for now and re-do it later on if necessary.

Reply:
What had only one pointer's worth of space? All allocations are bumped up to the nearest 16 bytes anyway, so a 40-byte object actually gets allocated as 48

---
In reply to: [205622702](https://github.com/Microsoft/ChakraCore/pull/5348#discussion_r205622702) [](ancestors = 205622702)

Reply:
I believe it was specifically on 32bit one of the types was otherwise full... I don't recall which at the moment though.